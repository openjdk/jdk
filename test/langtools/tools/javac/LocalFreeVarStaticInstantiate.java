/*
 * @test /nodynamiccopyright/
 * @bug 8322882 8345953
 * @summary Disallow attempts to access a free variable proxy field from a static method
 * @compile/fail/ref=LocalFreeVarStaticInstantiate.out -XDrawDiagnostics LocalFreeVarStaticInstantiate.java
 */

class LocalFreeVarStaticInstantiate {

    // local class in method
    static void foo(Object there) {
        class Local {
            {
                there.hashCode();
            }

            static {
                new Local();    // can't get there from here
            }

            static Runnable r = () -> {
                new Local();    // can't get there from here
            };
        }
    }

    // local class in lambda
    static Runnable foo = () -> {
        Object there = "";
        class Local {
            {
                there.hashCode();
            }

            static {
                new Local();    // can't get there from here
            }

            static Runnable r = () -> {
                new Local();    // can't get there from here
            };
        }
    };

    // local class in switch
    static Object bar = switch (foo) {
        case Runnable r -> {
            Object there = "";
            class Local {
                {
                    there.hashCode();
                }

                static {
                    new Local();    // can't get there from here
                }

                static Runnable r = () -> {
                    new Local();    // can't get there from here
                };
            }
            yield r;
        }
    };

    // local class in instance init
    {
        Object there = "";
        class Local {
            {
                there.hashCode();
            }

            static {
                new Local();    // can't get there from here
            }

            static Runnable r = () -> {
                new Local();    // can't get there from here
            };
        }
    }

    // local class in static init
    static {
        Object there = "";
        class Local {
            {
                there.hashCode();
            }

            static {
                new Local();    // can't get there from here
            }

            static Runnable r = () -> {
                new Local();    // can't get there from here
            };
        }
    }
}
