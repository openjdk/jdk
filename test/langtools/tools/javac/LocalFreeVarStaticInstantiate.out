LocalFreeVarStaticInstantiate.java:18:17: compiler.err.local.cant.be.inst.static: kindname.class, Local
LocalFreeVarStaticInstantiate.java:22:17: compiler.err.local.cant.be.inst.static: kindname.class, Local
LocalFreeVarStaticInstantiate.java:36:17: compiler.err.local.cant.be.inst.static: kindname.class, Local
LocalFreeVarStaticInstantiate.java:40:17: compiler.err.local.cant.be.inst.static: kindname.class, Local
LocalFreeVarStaticInstantiate.java:55:21: compiler.err.local.cant.be.inst.static: kindname.class, Local
LocalFreeVarStaticInstantiate.java:59:21: compiler.err.local.cant.be.inst.static: kindname.class, Local
LocalFreeVarStaticInstantiate.java:75:17: compiler.err.local.cant.be.inst.static: kindname.class, Local
LocalFreeVarStaticInstantiate.java:79:17: compiler.err.local.cant.be.inst.static: kindname.class, Local
LocalFreeVarStaticInstantiate.java:93:17: compiler.err.local.cant.be.inst.static: kindname.class, Local
LocalFreeVarStaticInstantiate.java:97:17: compiler.err.local.cant.be.inst.static: kindname.class, Local
10 errors
